File: Python/Problems/Problems/problem1759.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem


class Problem1759(Problem):
    def test(self) -> bool:
        s = "abbcccaa"

        expected = 13

        result = self.countHomogenous(s)

        print(result)

        return result == expected

    def countHomogenous(self, s: str) -> int:
        n = len(s)
        first = 0
        result = 0
        mod = 1000000007
        for i in range(n):
            if s[i] != s[first]:
                l = i - first
                result = (result + l * (l + 1) // 2) % mod
                first = i
        ls = n - first
        result = (result + ls * (ls + 1) // 2) % mod
        return result

File: C++/Problems/Problems/problem_1759.cpp
#include "problem_1759.h"

bool problem_1759::test() {
    std::string s = "abbcccaa";

    int expected = 13;

    auto result = countHomogenous(s);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_1759::countHomogenous(std::string s) {
    int n = s.size();
    int first = 0;
    long long result = 0;
    int mod = 1000000007;
    for (int i = 0; i < n; ++i)
    {
        if (s[i] != s[first])
        {
            long long l = i - first;
            result = (result + l * (l + 1) / 2) % mod;
            first = i;
        }
    }
    long long ls = n - first;
    result = (result + ls * (ls + 1) / 2) % mod;
    return (int)result;
}
